File: algorithms/medium/subarray_sum_equals_k.ex
defmodule Solution do
  @spec subarray_sum(nums :: [integer], k :: integer) :: integer
  def subarray_sum(nums, k) do
    nums
    |> Enum.reduce({0, 0, %{0 => 1}}, fn n, {ans, pre_sum, map} ->
      pre_sum = pre_sum + n
      ans = ans + Map.get(map, pre_sum - k, 0)
      {ans, pre_sum, Map.update(map, pre_sum, 1, &(&1 + 1))}
    end)
    |> elem(0)
  end
end

File: algorithms/medium/subarray_sum_equals_k.rs
use std::collections::HashMap;

impl Solution {
    pub fn subarray_sum(nums: Vec<i32>, k: i32) -> i32 {
        let mut map = HashMap::new();
        let mut pre_sum = 0;
        let mut ans = 0;
        map.insert(0, 1);
        for n in nums {
            pre_sum += n;
            if let Some(ct) = map.get(&(pre_sum - k)) {
                ans += *ct;
            }
            *map.entry(pre_sum).or_insert(0) += 1;
        }
        ans
    }
}
